### ISSUE TITLE
[no-unused-vars] does not detect name shadowing in type definition

### ISSUE THREAD

**Issue by miguel-leon** [typescript-eslint/typescript-eslint#3790]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
  }
}
```

```TS
type Tail<T extends any[]> = T extends [unknown, ...infer T] ? T : never;
```

```JSON
{
  "compilerOptions": {
    "target": "ESNext",
    "module": "ESNext",
    "strict": true,
    "outDir": "./dist",
    "declaration": true,
    "esModuleInterop": true,
    "noImplicitReturns": true,
    "moduleResolution": "Node"
  },
  "include": [
    "./src"
  ]
}
```

**Expected Result**

No warnings or errors.

**Actual Result**

```
warning  'T' is defined but never used        @typescript-eslint/no-unused-vars
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.29.3` |
| `@typescript-eslint/parser`        | `4.29.3` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.32.0` |
